commons/zenoh-buffers/src/zslice.rs

```diff
@@ -108,58 +97,52 @@ impl ZSlice {
         Self::new(buf, start, end)
     }
 
+    #[inline]
     pub fn new(
         buf: Arc<dyn ZSliceBuffer>,
         start: usize,
         end: usize,
     ) -> Result<ZSlice, Arc<dyn ZSliceBuffer>> {
         if start <= end && end <= buf.as_slice().len() {
-            // unsafe: this operation is safe because we just checked the slice boundaries
-            Ok(unsafe { ZSlice::new_unchecked(buf, start, end) })
+            Ok(Self { buf, start, end })
         } else {
             Err(buf)
         }
     }
 
+    #[inline]
     pub fn empty() -> Self {
-        Self::new(Arc::new([]), 0, 0).unwrap()
-    }
-
-    /// # Safety
-    /// This function does not verify whether the `start` and `end` indexes are within the buffer boundaries.
-    /// If a [`ZSlice`] is built via this constructor, a later access may panic if `start` and `end` indexes are out-of-bound.
-    pub unsafe fn new_unchecked(buf: Arc<dyn ZSliceBuffer>, start: usize, end: usize) -> Self {
-        ZSlice {
-            buf,
-            start,
-            end,
-            #[cfg(feature = "shared-memory")]
-            kind: ZSliceKind::Raw,
-        }
+        Self::new(Arc::new(Vec::<u8>::new()), 0, 0).unwrap()
     }
 
     #[inline]
     #[must_use]
-    pub fn downcast_ref<T>(&self) -> Option<&T>
-    where
-        T: Any,
-    {
-        self.buf.as_any().downcast_ref::<T>()
+    pub fn downcast_ref<T: Any>(&self) -> Option<&T> {
+        self.buf.as_any().downcast_ref()
     }
 
+    /// # Safety
+    ///
+    /// Buffer modification must not modify slice range.
     #[inline]
     #[must_use]
-    pub fn downcast_mut<T>(&mut self) -> Option<&mut T>
-    where
-        T: Any,
-    {
-        Arc::get_mut(&mut self.buf).and_then(|val| val.as_any_mut().downcast_mut::<T>())
+    pub unsafe fn downcast_mut<T: Any>(&mut self) -> Option<&mut T> {
```

Comment:
Why marking this function as `unsafe`? `Arc::get_mut` should already guarantee that reference count is 1 and it is safe to mutate.

Reply:
(This change is brought by https://github.com/eclipse-zenoh/zenoh/pull/1289 and should be discussed there)
The answer is in the safety comment. It may be safe to mutate the inner buffer inside the Arc, but `ZSlice` still have an invariant to maintain. If the function was safe, you would be able to truncate the vector, and the `end` would then be invalid. Thus, we need to ensure that the invariant is maintained, i.e. `start`/`end` remain valid for the modified slice. The simple precondition for that is to have the slice size unmodified, as it is the case for SHM buffers which are the only use case of this method.